apps/src/lib/node/ledger/protocol/mod.rs

```diff
@@ -95,22 +93,58 @@ where
 
 pub type Result<T> = std::result::Result<T, Error>;
 
-/// Apply a given transaction
+/// Dispatch a given transaction to be applied based on its type. Some storage
+/// updates may be derived and applied natively rather than via the wasm
+/// environment, in which case validity predicates will be bypassed.
 ///
 /// If the given tx is a successfully decrypted payload apply the necessary
 /// vps. Otherwise, we include the tx on chain with the gas charge added
 /// but no further validations.
-pub(crate) fn apply_tx<'a, D, H, CA>(
-    tx: TxType,
+pub(crate) fn dispatch_tx<'a, D, H, CA>(
+    tx_type: TxType,
     tx_length: usize,
-    ShellParams {
-        block_gas_meter,
-        write_log,
-        storage,
-        wasm_dir: _wasm_dir,
-        vp_wasm_cache,
-        tx_wasm_cache,
-    }: ShellParams<'a, D, H, CA>,
+    block_gas_meter: &'a mut BlockGasMeter,
+    write_log: &'a mut WriteLog,
+    storage: &'a mut Storage<D, H>,
+    vp_wasm_cache: &'a mut VpCache<CA>,
+    tx_wasm_cache: &'a mut TxCache<CA>,
+) -> Result<TxResult>
+where
+    D: 'static + DB + for<'iter> DBIter<'iter> + Sync,
+    H: 'static + StorageHasher + Sync,
+    CA: 'static + WasmCacheAccess + Sync,
+{
+    match tx_type {
+        TxType::Raw(_) => Err(Error::TxTypeError),
+        TxType::Decrypted(DecryptedTx::Decrypted(tx)) => apply_wasm_tx(
+            tx,
+            tx_length,
+            block_gas_meter,
+            write_log,
+            storage,
+            vp_wasm_cache,
+            tx_wasm_cache,
+        ),
+        TxType::Protocol(ProtocolTx { tx, .. }) => {
+            apply_protocol_tx(tx, storage)
+        }
+        _ => {
+            // other transaction types we treat as a noop
+            Ok(TxResult::default())
+        }
```

Comment:
I think here we should return `Err(Error::TxTypeError)` rather than noop. The only transaction types we should ever be applying for Namada are decrypted transactions and protocol transactions.